src/iterator.cr

```diff
@@ -1377,4 +1376,96 @@ module Iterator(T)
       self
     end
   end
+
+  # Returns an iterator for each chunked elements where each chunk
+  # ends right **before** the given block's value is _truthy_.
+  #
+  # For example, to get chunks that end just before each uppercase letter:
+  #
+  # ```
+  # ary = ['a', 'b', 'C', 'd', 'E', 'F', 'g', 'h']
+  # #              ^         ^    ^
+  # iter = ary.slice_before(&.uppercase?)
+  # iter.next # => ['a', 'b']
+  # iter.next # => ['C', 'd']
+  # iter.next # => ['E']
+  # iter.next # => ['F', 'g', 'h']
+  # iter.next # => Iterator::Stop
+  # ```
+  #
+  # By default, a new array is created and yielded for each slice when invoking `next`.
+  # * If *reuse* is `false`, the method will create a new array for each chunk
+  # * If *reuse* is `true`, the method will create a new array and reuse it.
+  # * If *reuse* is an `Array`, that array will be reused
+  #
+  # This can be used to prevent many memory allocations when each slice of
+  # interest is to be used in a read-only fashion.
+  def slice_before(reuse : Bool | Array(T) = false, &block : T -> B) forall B
+    SliceBefore(typeof(self), T, B).new(self, block, reuse)
+  end
+
+  # :nodoc:
+  class SliceBefore(I, T, B)
+    include Iterator(Array(T))
+
+    @has_value_to_add = false
+    @value_to_add : T?
+
+    def initialize(@iterator : I, @block : T -> B, reuse)
+      @end = false
+
+      if reuse
+        if reuse.is_a?(Array)
+          @values = reuse
+        else
+          @values = [] of T
+        end
+        @reuse = true
+      else
+        @values = [] of T
+        @reuse = false
+      end
+    end
+
+    def next
+      return stop if @end
+
+      if @has_value_to_add
+        @has_value_to_add = false
+        @values.clear
+        @values << @value_to_add.as(T)
+        @value_to_add = nil
+      end
+
+      while true
+        value = @iterator.next
+
+        if value.is_a?(Stop)
+          @end = true
+          if @values.empty?
+            return stop
+          else
+            return @reuse ? @values : @values.dup
+          end
+        end
+
+        if @block.call(value) && !@values.empty?
```

Comment:
`!@values.empty?` is cheaper and should go before the `@block.call`. The optimizer won't be able to reorder as it won't be able to prove that `@block` doesn't modify `@values`.

Reply:
Good idea!